presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public abstract class AbstractExpressionRewriteRuleSet
+{
+    abstract Expression rewrite(Expression expression, Rule.Context context);
+
+    public Set<Rule> getRules()
+    {
+        return ImmutableSet.of(
+                new ProjectExpressionRewrite(),
+                new AggregationExpressionRewrite(),
+                new FilterExpressionRewrite(),
+                new TableScanExpressionRewrite(),
+                new JoinExpressionRewrite(),
+                new ValuesExpressionRewrite(),
+                new ApplyExpressionRewrite()
+        );
+    }
+
+    private static class Patterns
+    {
+        private static final Pattern PROJECT = Pattern.typeOf(ProjectNode.class);
+        private static final Pattern AGGREGATION = Pattern.typeOf(AggregationNode.class);
+        private static final Pattern FILTER = Pattern.typeOf(FilterNode.class);
+        private static final Pattern TABLE_SCAN = Pattern.typeOf(TableScanNode.class);
+        private static final Pattern JOIN = Pattern.typeOf(JoinNode.class);
+        private static final Pattern VALUES = Pattern.typeOf(ValuesNode.class);
+        private static final Pattern APPLY = Pattern.typeOf(ApplyNode.class);
+    }
+
+    final class ProjectExpressionRewrite
```

Comment:
Left them package private for testing. I can use `@VisibleForTesting` but then I need to define constructors explicitly. Do you prefer this way?